package.json

```
@@ -1,10 +1,10 @@
 {
   "scripts": {
-    "start": "yarn start:firestore & yarn start:dev",
+    "start": "run-p start:firestore start:dev",
     "start:dev": "snowpack dev",
     "start:firestore": "firebase emulators:start --import=./devData --only firestore",
     "build": "snowpack build",
-    "test": "jest --watch",
+    "test": "jest --watch --FIRESTORE_EMULATOR_HOST=8080",
```

> when i started testing my app (which is also) using firebase. I used to get my API like this [...API_KEY]. so how did dotenv.config(); fix it? ... i used to think that Kent did something so people like me wouldn't do the mistake of testing using the same backend the app uses.

Dotenv just imported whatever I had in my `.env` file, as without those, it would fall over. This also meant that I don't have to worry about var being added or removed. It becomes extremely(!!!!) important to make sure you aren't "leaking" anything out to the real world, however. In Scrivener, I am managing this by using MSW with a catchall:
```typescript
  rest.get('*', (req, res, ctx) => {
    const warning = `${req.url} has not been mocked yet... Maybe now is a good time to do that`;
    console.warn(warning);
    return res(ctx.json(warning));
  }),
```
And then firestore has this really nice feature of the `FIRESTORE_EMULATOR_HOST` env var, which will make it point to the emulator instead of real firebase.

---

> Sorry, this would NOT work. run-s runs things sequentially, and the firestore emulator just runs until stopped - so it would never run the test script.

I thought it will run firestore emulator then jest. but even run-p didn't work even thu it worked with `start` command.

> components/loadingScreen/loadingScreen.test.ts <-- this will place it right next to the file it is testing.
this will surely be a lot cleaner. I will do the changes.
@JacobMGEvans is it okay to do so?

---

i see now. that cleared alot of things for me. 🙏🏼 thanks  🙂 
using the emulator is way better than using MSW to fake the requests. because `firebase url` are pretty damn big.
```suggestion
    "test": "jest --watch",
```
I dont think that the `FIRESTORE_EMULATOR_HOST` var works here in this line. when i added it, it was for the test env to pick up on it but the way you wrote the script and configured the env var bypassed its use case.

---

@AhmedEldessouki the way you set it here doesn't set an env var. You will have to use something like cross-env and set it _first_ - see Guild Scrivener for how I setup a generic "execute firestore" command that everything else hooks off of for a solid example.

---

thanks for pointing that out 🙂 I will look for it 😁😁

---

Cant work around this. int the Guild Scrivener you are using Node env. do you think it's possible for `@firebase/testing` to work in Jsdom env?
`Error: FIRESTORE (8.2.4) INTERNAL ASSERTION FAILED: Unexpected state`

I ran into this problem before when i was testing firestore-rules (Node env). after i worked around it, jsdom tests was being unexpected and then i removed firestore-rules test